core/lib/server/pub-filters.db.test.ts

```diff
@@ -0,0 +1,961 @@
+			values: {
+				Vector3: [0, 0, 0],
+			},
+		},
+		{
+			id: numberArrayId,
+			pubType: "Basic Pub",
+			values: {
+				NumberArray: [1, 2, 3],
+				Date: twenty99,
+			},
+		},
+		{
+			id: numberArray2Id,
+			pubType: "Basic Pub",
+			values: {
+				NumberArray: [10, 20, 30, 40],
+			},
+		},
+		{
+			id: relationId,
+			pubType: "Basic Pub",
+			values: {
+				Relation: null,
+				Number: 50,
+			},
+		},
+		{
+			id: testTitleId,
+			pubType: "Basic Pub",
+			values: {
+				Title: "Test",
+				Number: 99,
+			},
+		},
+		{
+			id: testCaseId,
+			pubType: "Basic Pub",
+			values: {
+				Date: new Date("2023-02-01T00:00:00.000Z"),
+				Boolean: true,
+				Title: "Test case",
+			},
+		},
+		{
+			id: specialCharsId,
+			pubType: "Basic Pub",
+			values: {
+				Title: "Some title with special characters & symbols",
+			},
+		},
+		{
+			id: importantDocId,
+			pubType: "Basic Pub",
+			values: {
+				Title: "Test important document",
+				Number: 40,
+				Date: new Date("2023-02-01T00:00:00.000Z"),
+				Boolean: true,
+			},
+		},
+	],
+});
+
+// let community: CommunitySeedOutput<typeof seed>;
+
+const seedCommunity = async (trx = testDb) => {
+	const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
+	const community = await seedCommunity(
+		seed,
+		{
+			// no random slug like usual because we have to define `.each` tests statically
+			randomSlug: false,
+		},
+		trx
+	);
+
+	// set the updatedAt for a pub to a weird date
+	const p = await trx
+		.updateTable("pubs")
+		.set({ createdAt: new Date("2024-01-01") })
+		.where("id", "=", trueId)
+		.returningAll()
+		.execute();
+
+	await trx
+		.updateTable("pubs")
+		.set({ createdAt: twenty99 })
+		.where("id", "=", vector3Id)
+		.execute();
+
+	return community;
+};
+
+const community = await seedCommunity();
+
+// afterAll(async () => {
+// 	const { deleteCommunity } = await import("~/prisma/seed/deleteCommunity");
+// 	await deleteCommunity(communityId);
+// });
+
+const coolQuery = (filter: Filter) => {
+	const trx = getTrx();
+
+	const q = trx
+		.selectFrom("pubs")
+		.selectAll()
+		.where((eb) => applyFilters(eb, filter));
+
+	return {
+		q,
+		compiled: q.compile(),
+	};
+};
+
+const validateFilter = async (communityId: CommunitiesId, filter: Filter, trx = getTrx()) => {
+	const { validateFilter: valFilter } = await import("./validate-filters");
+	return valFilter(communityId, filter, trx);
+};
+
+const slug = (str: string) => `${communitySlug}:${str}`;
+
+const unifiedTestCases: {
+	title: string;
+	filter: Filter;
+	querystring: string;
+	sql: string | RegExp | (string | RegExp)[];
+	parameters: (string | number)[];
+	foundIds: PubsId[] | ((pubs: ProcessedPub[]) => PubsId[]);
+}[] = [
+	{
+		title: "simple equality",
+		filter: {
+			[slug("title")]: { $eq: "Some title" },
+		},
+		querystring: `filters[${slug("title")}][$eq]=Some title`,
+		sql: `"slug" = $1 and "value" = $2`,
+		parameters: [slug("title"), '"Some title"'],
+		foundIds: [titleId],
+	},
```

Comment:
im running all these test cases through a number of tests, namely

- checking whether the querystring gets converted to the proper filter object
- whether said filter object passes parsing
- whether said filter object generates the correct `sql`
- whether said filter ends up fetching the correct pubs